File: solutions/005-longest-palindromic-substring/longest-palindromic-substring.py
# -*- coding:utf-8 -*-


# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
#
# Example 1:
#
#
# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.
#
#
# Example 2:
#
#
# Input: "cbbd"
# Output: "bb"
#
#


class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        longest, mid = "", (len(s) - 1) / 2
        i, j = mid, mid
        while i >= 0 and j < len(s):
            args = [(s, i, i), (s, i, i + 1), (s, j, j), (s, j, j + 1)]
            for arg in args:
                tmp = self.longestPalindromeByAxis(*arg)
                if len(tmp) > len(longest):
                    longest = tmp
            if len(longest) >= i * 2:
                if len(longest) == 1:
                    return s[0]
                return longest
            i, j = i - 1, j + 1
        return longest

    def longestPalindromeByAxis(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left, right = left - 1, right + 1
        return s[left + 1: right]

File: solutions/005-longest-palindromic-substring/longest-palindromic-substring.cpp
// Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
//
// Example 1:
//
//
// Input: "babad"
// Output: "bab"
// Note: "aba" is also a valid answer.
//
//
// Example 2:
//
//
// Input: "cbbd"
// Output: "bb"
//
//


class Solution {
public:
    string preManache(string s) {
        string n_s = "!";
        for (auto c : s) {
            n_s += '#';
            n_s += c;
        }
        n_s += "#?";
        return n_s;
    }
    string longestPalindrome(string s) {
        if (s.empty()) {
            return s;
        }
        string n_s = preManache(s);
        int size = n_s.size();
        vector<int> radium(size, 0);
        int right = 0, mid = 0, max_r = 0, p = 0, pivot = 0;
        for (int i = 1; i < size - 1; ++i) {
            if (i < right) {
                radium[i] = min(right - i, radium[2*mid - i]);
            }
            while (n_s[i - radium[i] - 1] == n_s[i + radium[i] + 1]) {
                ++radium[i];
            }
            if (i + radium[i] > right) {
                right = i + radium[i];
                mid = i;
            }
            if (radium[i] > max_r) {
                max_r = radium[i];
                pivot = i;
                p = (pivot - max_r - 1) / 2;
            }
        }
        return s.substr(p, max_r);
    }
};
